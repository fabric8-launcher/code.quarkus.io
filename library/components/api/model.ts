
export interface QuarkusProject {
  metadata: {
    groupId: string;
    artifactId: string;
    version: string;
    name?: string;
    noCode?: boolean;
    buildTool: string;
    javaVersion: string;
  };
  extensions: string[];
  streamKey?: string;
  platformOnly?: boolean;
  github?: {
    code: string;
    state: string;
  };
}

export interface Tag {
  name: string;
  href?: string;
  description?: string;
  color?: string;
  hide?: boolean;
}

export interface Extension {
  id: string;
  version: string;
  name: string;
  keywords: string[];
  tags: string[];
  description?: string;
  shortName?: string;
  category: string;
  platform: boolean;
  default: boolean;
  order: number;
  guide?: string;
  bom?: string;
}

export interface PlatformMappedExtensions {
  mapped: Extension[];
  missing: string[];
}

export interface Platform {
  extensions: Extension[];
  streams: Stream[];
  tagsDef: Tag[];
}

export interface Stream {
  key: string;
  quarkusCoreVersion: string;
  platformVersion: string;
  recommended: boolean;
  status: string;
}

export interface Config {
  environment: string;
  segmentWriteKey?: string;
  sentryDSN?: string;
  quarkusPlatformVersion: string;
  gitCommitId: string;
  gitHubClientId?: string;
  features: string[];
}


