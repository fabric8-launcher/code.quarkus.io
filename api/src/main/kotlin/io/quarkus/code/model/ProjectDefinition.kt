package io.quarkus.code.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import org.eclipse.microprofile.openapi.annotations.media.Schema
import org.eclipse.microprofile.openapi.annotations.parameters.Parameter
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.Pattern
import javax.ws.rs.DefaultValue
import javax.ws.rs.QueryParam

@JsonIgnoreProperties(ignoreUnknown = true)
data class ProjectDefinition(val streamKey: String? = null,
                             val groupId: String = DEFAULT_GROUPID,
                             val artifactId: String = DEFAULT_ARTIFACTID,
                             val version: String = DEFAULT_VERSION,
                             val className: String? = null,
                             val path: String? = null,
                             val buildTool: String = DEFAULT_BUILDTOOL,
                             val javaVersion: String = DEFAULT_JAVA_VERSION,
                             val noCode: Boolean = false,
                             val extensions: Set<String> = setOf()) {

    companion object {
        const val DEFAULT_GROUPID = "org.acme"
        const val DEFAULT_ARTIFACTID = "code-with-quarkus"
        const val DEFAULT_VERSION = "1.0.0-SNAPSHOT"
        const val DEFAULT_BUILDTOOL = "MAVEN"
        const val DEFAULT_NO_CODE = false
        const val DEFAULT_JAVA_VERSION = "17"

        const val JAVA_VERSION_PATTERN = "^(?:1\\.)?(\\d+)(?:\\..*)?\$";
        const val GROUPID_PATTERN = "^([a-zA-Z_\$][a-zA-Z\\d_\$]*\\.)*[a-zA-Z_\$][a-zA-Z\\d_\$]*\$"
        const val ARTIFACTID_PATTERN = "^[a-z][a-z0-9-._]*\$"
        const val CLASSNAME_PATTERN = GROUPID_PATTERN
        const val PATH_PATTERN = "^\\/([a-z0-9\\-._~%!\$&'()*+,;=:@]+\\/?)*\$"
        const val BUILDTOOL_PATTERN = "^(MAVEN)|(GRADLE)|(GRADLE_KOTLIN_DSL)\$"
    }
}