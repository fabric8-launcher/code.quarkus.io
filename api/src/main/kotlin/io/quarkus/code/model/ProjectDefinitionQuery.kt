package io.quarkus.code.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import io.quarkus.code.model.ProjectDefinition.Companion.ARTIFACTID_PATTERN
import io.quarkus.code.model.ProjectDefinition.Companion.BUILDTOOL_PATTERN
import io.quarkus.code.model.ProjectDefinition.Companion.CLASSNAME_PATTERN
import io.quarkus.code.model.ProjectDefinition.Companion.DEFAULT_ARTIFACTID
import io.quarkus.code.model.ProjectDefinition.Companion.DEFAULT_BUILDTOOL
import io.quarkus.code.model.ProjectDefinition.Companion.DEFAULT_GROUPID
import io.quarkus.code.model.ProjectDefinition.Companion.DEFAULT_JAVA_VERSION
import io.quarkus.code.model.ProjectDefinition.Companion.DEFAULT_NO_CODE
import io.quarkus.code.model.ProjectDefinition.Companion.DEFAULT_VERSION
import io.quarkus.code.model.ProjectDefinition.Companion.GROUPID_PATTERN
import io.quarkus.code.model.ProjectDefinition.Companion.JAVA_VERSION_PATTERN
import io.quarkus.code.model.ProjectDefinition.Companion.PATH_PATTERN
import org.eclipse.microprofile.openapi.annotations.media.Schema
import org.eclipse.microprofile.openapi.annotations.parameters.Parameter
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.Pattern
import javax.ws.rs.DefaultValue
import javax.ws.rs.QueryParam

@JsonIgnoreProperties(ignoreUnknown = true)
class ProjectDefinitionQuery {

    @QueryParam("S")
    @Parameter(name = "S", description = "The platform stream to use to create this project ('platformKey:streamId' or 'streamId')", required = false)
    @Schema(description = "The platform stream to use to create this project ('platformKey:streamId' or 'streamId')", required = false)
    var streamKey: String? = null
        private set

    @DefaultValue(DEFAULT_GROUPID)
    @NotEmpty
    @Pattern(regexp = GROUPID_PATTERN)
    @QueryParam("g")
    @Parameter(name = "g", description = "GAV: groupId", required = false)
    @Schema(description = "GAV: groupId", required = false, defaultValue = DEFAULT_GROUPID, pattern = GROUPID_PATTERN)
    var groupId: String = DEFAULT_GROUPID
        private set

    @DefaultValue(DEFAULT_ARTIFACTID)
    @NotEmpty
    @Pattern(regexp = ARTIFACTID_PATTERN)
    @QueryParam("a")
    @Parameter(name = "a", description = "GAV: artifactId", required = false)
    @Schema(description = "GAV: artifactId", required = false, defaultValue = DEFAULT_ARTIFACTID, pattern = ARTIFACTID_PATTERN)
    var artifactId: String = DEFAULT_ARTIFACTID
        private set

    @DefaultValue(DEFAULT_VERSION)
    @NotEmpty
    @QueryParam("v")
    @Parameter(name = "v", description = "GAV: version", required = false)
    @Schema(description = "GAV: version", required = false, defaultValue = DEFAULT_VERSION)
    var version: String = DEFAULT_VERSION
        private set

    @QueryParam("c")
    @Pattern(regexp = CLASSNAME_PATTERN)
    @Parameter(name = "c", description = "The class name to use in the generated application", required = false)
    @Schema(description = "The class name to use in the generated application", required = false, pattern = CLASSNAME_PATTERN)
    var className: String? = null
        private set

    @QueryParam("p")
    @Pattern(regexp = PATH_PATTERN)
    @Parameter(name = "p", description = "The path of the REST endpoint created in the generated application", required = false)
    @Schema(description = "The path of the REST endpoint created in the generated application", required = false, pattern = PATH_PATTERN)
    var path: String? = null
        private set

    @DefaultValue(DEFAULT_NO_CODE.toString())
    @QueryParam("ne")
    @Parameter(name = "ne", description = "No code examples (Deprecated: use noCode (nc) instead)", required = false, schema = Schema(deprecated = true))
    @Schema(description = "No code examples (Deprecated: use noCode (nc) instead)", deprecated = true, required = false, defaultValue = DEFAULT_NO_CODE.toString())
    @Deprecated(message = "Use noCode (nc) instead")
    var noExamples: Boolean = DEFAULT_NO_CODE
        private set

    @DefaultValue(DEFAULT_NO_CODE.toString())
    @QueryParam("nc")
    @Parameter(name = "nc", description = "No code", required = false)
    @Schema(description = "No code", required = false, defaultValue = DEFAULT_NO_CODE.toString())
        var noCode: Boolean = DEFAULT_NO_CODE
        private set

    @DefaultValue(DEFAULT_BUILDTOOL)
    @NotEmpty
    @QueryParam("b")
    @Pattern(regexp = BUILDTOOL_PATTERN)
    @Parameter(name = "b", description = "The build tool to use (MAVEN, GRADLE or GRADLE_KOTLIN_DSL)", required = false, schema = Schema(enumeration = ["MAVEN", "GRADLE", "GRADLE_KOTLIN_DSL"]))
    @Schema(description = "The build tool to use (MAVEN, GRADLE or GRADLE_KOTLIN_DSL)", enumeration = ["MAVEN", "GRADLE", "GRADLE_KOTLIN_DSL"], defaultValue = DEFAULT_BUILDTOOL)
    var buildTool: String = DEFAULT_BUILDTOOL
        private set

    @QueryParam("j")
    @NotEmpty
    @DefaultValue(DEFAULT_JAVA_VERSION)
    @Pattern(regexp = JAVA_VERSION_PATTERN)
    @Parameter(name = "j", description = "The Java version for the generation application", required = false)
    @Schema(description = "The Java version for the generation application", required = false, pattern = JAVA_VERSION_PATTERN)
    var javaVersion: String = DEFAULT_JAVA_VERSION
        private set


    @QueryParam("e")
    @Parameter(name = "e", description = "The set of extension ids that will be included in the generated application", required = false)
    @Schema(description = "The set of extension ids that will be included in the generated application", required = false)
    var extensions: Set<String> = setOf()
        private set

    fun toProjectDefinition(): ProjectDefinition {
        return ProjectDefinition(
            streamKey = streamKey,
            groupId = groupId,
            artifactId = artifactId,
            version = version,
            className = className,
            path = path,
            buildTool = buildTool,
            javaVersion = javaVersion,
            noCode = noCode || noExamples,
            extensions = extensions
        )
    }
}