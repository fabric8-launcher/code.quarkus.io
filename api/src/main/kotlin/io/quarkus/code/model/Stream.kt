package io.quarkus.code.model

data class Stream(
    val key: String,
    val quarkusCoreVersion: String,
    val platformVersion: String,
    val recommended: Boolean,
    val status: String
)
