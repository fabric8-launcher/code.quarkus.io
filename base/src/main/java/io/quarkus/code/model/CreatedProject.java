package io.quarkus.code.model;

public record CreatedProject(String path) {}