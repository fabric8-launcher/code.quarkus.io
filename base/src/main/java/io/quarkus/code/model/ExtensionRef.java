package io.quarkus.code.model;

public record ExtensionRef(
        String id,
        String version,
        boolean platform) {

}
