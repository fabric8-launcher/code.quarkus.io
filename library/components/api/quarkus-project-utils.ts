import { parse, ParsedUrlQuery, stringify } from 'querystring';
import { createGitHubProject } from './code-quarkus-github-api';
import { Extension, PlatformMappedExtensions, QuarkusProject } from './model';
import _ from 'lodash';
import { Api } from './code-quarkus-api';

export enum Target {
  GENERATE = 'GENERATE',
  DOWNLOAD = 'DOWNLOAD',
  GITHUB = 'GITHUB'
}

export enum LocalStorageKey {
  DEFAULT_PROJECT = 'quarkus:default-project-config'
}

export interface GenerateResult {
  target: Target;
  url: string;
  shareUrl?: string;
}

export interface ProjectPayload {
  groupId?: string;
  artifactId?: string;
  version?: string;
  buildTool?: string;
  javaVersion?: string;
  noCode?: boolean;
  extensions?: string[];
  platformOnly?: boolean;
  streamKey?: string;
}

export function generateProjectPayload(project: QuarkusProject): ProjectPayload {
  const defaultProject = newDefaultProject();
  const changes: ProjectPayload = {};
  if (project.metadata.groupId && project.metadata.groupId !== defaultProject.metadata.groupId) {
    changes.groupId = project.metadata.groupId;
  }
  if (project.metadata.artifactId && project.metadata.artifactId !== defaultProject.metadata.artifactId) {
    changes.artifactId = project.metadata.artifactId;
  }
  if (project.metadata.version && project.metadata.version !== defaultProject.metadata.version) {
    changes.version = project.metadata.version;
  }
  if (project.metadata.buildTool && project.metadata.buildTool !== defaultProject.metadata.buildTool) {
    changes.buildTool = project.metadata.buildTool;
  }
  if (project.metadata.javaVersion && project.metadata.javaVersion !== defaultProject.metadata.javaVersion) {
    changes.javaVersion = project.metadata.javaVersion;
  }
  if (project.metadata.noCode && project.metadata.noCode !== defaultProject.metadata.noCode) {
    changes.noCode = project.metadata.noCode;
  }
  if (project.streamKey) {
    changes.streamKey = project.streamKey;
  }
  if (project.platformOnly !== undefined) {
    changes.platformOnly = project.platformOnly;
  }
  if (project.extensions && project.extensions.length > 0) {
    changes.extensions = project.extensions;
  }
  return changes;
}

const QUERY_PROJECT_MAPPING: any = {
  groupId: 'g',
  artifactId: 'a',
  version: 'v',
  buildTool: 'b',
  javaVersion: 'j',
  noCode: 'nc',
  extensions: 'e',
  streamKey: 'S',
  platformOnly: 'po'
};

export function generateProjectQuery(api: Api, project: QuarkusProject,
  github: boolean = false,
  showClientName: boolean = true): string {
  const payload = generateProjectPayload(project);
  const params: any = {};
  for (const key in QUERY_PROJECT_MAPPING) {
    if (QUERY_PROJECT_MAPPING.hasOwnProperty(key)) {
      const pValue = (payload as any)[key];
      if (pValue !== undefined) {
        if (key === 'extensions') {
          params[QUERY_PROJECT_MAPPING['extensions']] = payload.extensions!!.map(id => toShortcut(id));
        } else {
          params[QUERY_PROJECT_MAPPING[key]] = pValue;
        }
      }
    }
  }
  if (showClientName) {
    params.cn = api.clientName;
  }

  if (github) {
    params.github = true;
  }

  return stringify(params);
}

export function toShortcut(id: string) {
  return id.replace(/^(io.quarkus:quarkus-)|^(quarkus-)/, '');
}

const BASE_LOCATION = window.location.href.replace(window.location.search, '');

export function getProjectDownloadUrl(api: Api, project: QuarkusProject) {
  const baseUrl = api.backendUrl.startsWith('http') ? api.backendUrl : BASE_LOCATION;
  return `${baseUrl.replace(/\/$/, '')}/d?${generateProjectQuery(api, project)}`;
}

export function getProjectShareUrl(api: Api, project: QuarkusProject, github = false) {
  return `${BASE_LOCATION}?${generateProjectQuery(api, project, github, false)}`;
}

export async function generateProject(api: Api, environment: string, project: QuarkusProject, target: Target): Promise<GenerateResult> {
  switch (target) {
  case Target.DOWNLOAD:
  case Target.GENERATE:
    const url = getProjectDownloadUrl(api, project);
    const shareUrl = getProjectShareUrl(api, project);
    if (target === Target.DOWNLOAD ) {
      setTimeout(() => window.open(url, '_blank'), 500);
    }
    return { target, url, shareUrl };
  case Target.GITHUB:
    const result = await createGitHubProject(api, project);
    return { target, url: result.url };
  }
}

export const createOnGitHub = (api: Api, project: QuarkusProject, clientId: string) => {
  const authParams = {
    redirect_uri: getProjectShareUrl(api, project, true),
    client_id: clientId,
    scope: 'public_repo,workflow',
    state: Math.random().toString(36)
  };
  const githubAuthorizeUrl = `https://github.com/login/oauth/authorize?${stringify(authParams)}`;
  window.location.href = githubAuthorizeUrl;
};

export function newDefaultProject(): QuarkusProject {
  return ({
    metadata: {
      groupId: 'org.acme',
      artifactId: 'code-with-quarkus',
      version: '1.0.0-SNAPSHOT',
      buildTool: 'MAVEN',
      javaVersion: '17',
      noCode: false
    },
    extensions: [],
  });
}

const FILTER_PARAM_NAME = 'extension-search';

function syncParamsInQuery(api: Api, project: QuarkusProject | undefined, filterParam: string = ''): void {
  const filter = filterParam.trim() === DEFAULT_FILTER.trim() ? '' : filterParam;
  if (!project) {
    window.history.replaceState(null, '', `/?${formatParam(FILTER_PARAM_NAME, filter)}`);
    return;
  }
  window.history.replaceState(null, '', `/${generateParamQuery(formatParam(FILTER_PARAM_NAME, filter), generateProjectQuery(api, project, false, false))}`);
}

export const debouncedSyncParamsQuery = _.debounce(syncParamsInQuery, 500);

const defaultCleanHistory = () => {
  console.log(`remove query from url: ${window.location.search}`);
  window.history.replaceState({}, document.title, window.location.href.replace(window.location.search, ''));
};

export function resolveQueryParams(search: string = window.location.search.substr(1),
  cleanHistory: () => void = defaultCleanHistory): ParsedUrlQuery | undefined {
  if (search.length === 0) {
    return undefined;
  }
  let queryParams = parse(search);
  cleanHistory();
  return queryParams;
}

let queryParams: ParsedUrlQuery | undefined = undefined;

export function normalizeStreamKey(recommendedPlatformId: string, streamKey?: string) {
  if (streamKey == null) {
    return null;
  }
  return streamKey!.indexOf(':') >= 0 ? streamKey! : `${recommendedPlatformId}:${streamKey}`;
}

export function getQueryParams(): ParsedUrlQuery | undefined {
  if (!queryParams) {
    queryParams = resolveQueryParams();
  }
  return queryParams;
}

export const DEFAULT_FILTER = 'origin:platform '

export function resolveInitialFilterQueryParam(queryParams = getQueryParams()): string {
  if (!queryParams || !queryParams[FILTER_PARAM_NAME]) {
    return DEFAULT_FILTER;
  }

  return queryParams![FILTER_PARAM_NAME]!.toString();
}

const formatParam = (paramName: string, value: string): string => {
  if (value) {
    return `${paramName}=${value}`;
  }

  return '';
};

const generateParamQuery = (filter: string, project: string) => {
  if (filter && project) {
    return `?${project}&${filter}`;
  }

  if (project) {
    return `?${project}`;
  }

  if (filter) {
    return `?${filter}`;
  }

  return '';
};

export function resolveInitialProject(queryParams?: ParsedUrlQuery) {
  return parseProjectInQuery(queryParams) || retrieveProjectFromLocalStorage() || newDefaultProject();
}

function normalizeQueryExtensions(queryExtensions: undefined | string | string[]): Set<string> {
  return new Set((queryExtensions ? Array.isArray(queryExtensions) ? queryExtensions : [ queryExtensions ] : [])
    .map(e => toShortcut(e)));
}

export function mapExtensions(catalog: Extension[], extensions: string[]): PlatformMappedExtensions {
  const selected = extensions.map(e => toShortcut(e));
  const catalogMap = new Map(catalog.map(e => [ toShortcut(e.id), e ]));
  const missing = extensions.filter(id => !catalogMap.has(toShortcut(id)))
  let mapped = selected.filter(id => catalogMap.has(id)).map(id => catalogMap.get(id));
  return  {
    mapped,
    missing
  };
}

export function parseProjectInQuery(queryParams?: ParsedUrlQuery): QuarkusProject | undefined {
  if (!queryParams) {
    return undefined;
  }
  const queryExtensions = normalizeQueryExtensions(queryParams?.e);
  const defaultProj = newDefaultProject();
  const project = {
    metadata: {
      groupId: queryParams.g || defaultProj.metadata.groupId,
      artifactId: queryParams.a || defaultProj.metadata.artifactId,
      version: queryParams.v || defaultProj.metadata.version,
      buildTool: queryParams.b || defaultProj.metadata.buildTool,
      javaVersion: queryParams.j || defaultProj.metadata.javaVersion,
      noCode: queryParams.nc || defaultProj.metadata.noCode
    },
    extensions: Array.from(queryExtensions),
    streamKey: queryParams.S,
    platformOnly: queryParams.po || defaultProj.platformOnly,
    github: queryParams.github === 'true' ? {
      code: queryParams.code,
      state: queryParams.state
    } : undefined
  } as QuarkusProject;
  if (project.github) {
    console.log('Received GitHub auth');
  }
  return project;
}

export function retrieveProjectFromLocalStorage() : QuarkusProject | undefined {
  try {
    const jsonProject = localStorage.getItem(LocalStorageKey.DEFAULT_PROJECT);

    if(!jsonProject) {
      return undefined;
    }

    const project = JSON.parse(jsonProject!) as QuarkusProject;
    if (!isValidQuarkusProject(project)) {
      return undefined;
    }
  
    return project;
  } catch (err) {
    return undefined;
  }
}

export function saveProjectToLocalStorage(project : QuarkusProject) {
  const jsonProject = JSON.stringify(project);
  localStorage.setItem(LocalStorageKey.DEFAULT_PROJECT, jsonProject);
}

//Basic validation (prevent changes on localstorage json, breaking project structure)
function isValidQuarkusProject(project : QuarkusProject): Boolean {
  const valid = project.metadata !== undefined && 
                project.metadata.groupId !== undefined &&
                project.metadata.artifactId !== undefined &&
                project.metadata.version !== undefined &&
                project.metadata.buildTool !== undefined &&
                project.metadata.javaVersion !== undefined &&
                project.extensions !== undefined;
  return valid;
}