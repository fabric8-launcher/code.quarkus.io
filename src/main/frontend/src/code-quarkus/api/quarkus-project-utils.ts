import { parse, stringify } from 'querystring';
import { ExtensionEntry } from '../pickers/extensions-picker';
import { createGitHubProject } from './code-quarkus-github-api';
import { QuarkusProject } from './model';
import { BACKEND_URL, CLIENT_NAME } from './env';

export enum Target {
  DOWNLOAD = 'DOWNLOAD',
  GITHUB = 'GITHUB'
}

export function generateProjectQuery(project: QuarkusProject, github: boolean = false): string {
  const packageName = project.metadata.packageName || project.metadata.groupId;
  const params: any = {
    ...(project.metadata.groupId && { g: project.metadata.groupId }),
    ...(project.metadata.artifactId && { a: project.metadata.artifactId }),
    ...(project.metadata.version && { v: project.metadata.version }),
    ...(project.metadata.buildTool && { b: project.metadata.buildTool }),
    ...(packageName && { c: `${packageName}.ExampleResource` }),
    ...(project.extensions && { s: project.extensions.map(e => e.shortId).join('.') }),
    cn: CLIENT_NAME
  };
  if (github) {
    params.github = true;
  }
  return stringify(params);
}

export function getProjectDownloadUrl(project: QuarkusProject) {
  return `${BACKEND_URL}/d?${generateProjectQuery(project)}`;
}


export async function generateProject(environment: string, project: QuarkusProject, target: Target): Promise<{ target: Target, url: string }> {
  switch (target) {
    case Target.DOWNLOAD:
      const url = getProjectDownloadUrl(project);
      if (environment !== 'dev') {
        setTimeout(() => window.open(url, '_blank'), 1000);
      }
      return { target, url };
    case Target.GITHUB:
      const result = await createGitHubProject(project);
      return { target, url: result.url };
  }
}

export const createOnGitHub = (project: QuarkusProject, clientId: string) => {
  const authParams = {
    redirect_uri: `${window.location.href.replace(window.location.search, '')}?${generateProjectQuery(project, true)}`,
    client_id: clientId,
    scope: 'repo',
    state: Math.random().toString(36)
  };
  const githubAuthorizeUrl = `https://github.com/login/oauth/authorize?${stringify(authParams)}`;
  window.location.href = githubAuthorizeUrl;
};

export function newDefaultProject(): QuarkusProject {
  return ({
    metadata: {
      groupId: 'org.acme',
      artifactId: 'code-with-quarkus',
      version: '1.0.0-SNAPSHOT',
      buildTool: 'MAVEN'
    },
    extensions: [],
  });
}

export function resolveInitialProject(extensions: ExtensionEntry[]) {
  return parseProjectInQuery(extensions) || newDefaultProject();
}

const defaultCleanHistory = () => {
  console.log(`remove query from url: ${window.location.search}`);
  window.history.replaceState({}, document.title, window.location.href.replace(window.location.search, ''));
};

export function parseProjectInQuery(extensions: ExtensionEntry[],
                                    search: string = window.location.search.substr(1),
                                    cleanHistory: () => void = defaultCleanHistory): QuarkusProject | undefined {
  if (search.length === 0) {
    return undefined;
  }
  const queryObj = parse(search);
  const shortIds = new Set((typeof queryObj.s === 'string') ? (queryObj.s as string).split('.') : []);
  const ids = new Set();
  const selectedExtensions = extensions.filter(e => {
    if (shortIds.has(e.shortId)) {
      const already = ids.has(e.id);
      ids.add(e.id);
      return !already;
    }
    return false;
  });
  const defaultProj = newDefaultProject();
  const project = {
    metadata: {
      groupId: queryObj.g || defaultProj.metadata.groupId,
      artifactId: queryObj.a || defaultProj.metadata.artifactId,
      version: queryObj.v || defaultProj.metadata.version,
      buildTool: queryObj.b || defaultProj.metadata.buildTool,
      packageName: (typeof queryObj.c === 'string') ? (queryObj.c as string).replace('.ExampleResource', '') : undefined,
    },
    extensions: selectedExtensions,
    github: queryObj.github === 'true' ? {
      code: queryObj.code,
      state: queryObj.state
    } : undefined
  } as QuarkusProject;
  if (project.github) {
    console.log('Received GitHub auth');
  }
  cleanHistory();
  return project;
}