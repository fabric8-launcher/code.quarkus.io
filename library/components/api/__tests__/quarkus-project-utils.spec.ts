import { cleanup } from '@testing-library/react';
import { mapExtensions, parseProjectInQuery, toShortcut, retrieveProjectFromLocalStorage, LocalStorageKey, saveProjectToLocalStorage, resetProjectToDefault, existsStoredProject, parse } from '../quarkus-project-utils';
import { Extension, QuarkusProject } from '../model';

beforeAll(()=> {
  configureLocalstorageMock();
});

afterEach(() => {
  cleanup();
});

function configureLocalstorageMock() {
  jest.spyOn(window.localStorage.__proto__, "setItem");
  jest.spyOn(window.localStorage.__proto__, "getItem");
  jest.spyOn(window.localStorage.__proto__, "removeItem");

  window.localStorage.__proto__.setItem = jest.fn();
  window.localStorage.__proto__.getItem = jest.fn();
  window.localStorage.__proto__.removeItem = jest.fn();
}

const entries: Extension[] = [
  {
    'id': 'io.quarkus:quarkus-arc',
    'name': 'ArC',
    'version': 'test-version',
    'tags': [],
    'keywords': [
      'arc',
      'cdi',
      'dependency-injection',
      'di',
      'label'
    ],
    'platform': true,
    'default': false,
    'description': 'Build time CDI dependency injection',
    'shortName': 'CDI',
    'category': 'Core',
    'order': 0,
  },
  {
    'id': 'io.quarkus:quarkus-camel-netty4-http',
    'version': 'test-version',
    'name': 'Camel Netty4 test HTTP',
    'tags': [ 'status:preview' ],
    'default': false,
    'keywords': [
      'camel-netty4-http',
      'camel'
    ],
    'platform': true,
    'description': 'Camel support for Netty',
    'category': 'Integration',
    'order': 2,
  },
  {
    'id': 'some-id',
    'version': 'test-version',
    'name': 'A CDI in name test',
    'tags': [ 'status:experimental' ],
    'default': false,
    'keywords': [
      'lambda',
      'amazon-lambda',
      'aws-lambda',
      'amazon',
      'aws',
      'label'
    ],
    'platform': true,
    'shortName': 'a shortname',
    'description': 'Some description',
    'category': 'Cloud',
    'order': 1,
  },
  {
    'id': 'some-id3',
    'version': 'test-version',
    'name': 'A CDI in name test',
    'tags': [ 'status:experimental' ],
    'default': false,
    'keywords': [
      'lambda',
      'amazon-lambda',
      'aws-lambda',
      'amazon',
      'aws',
      'label'
    ],
    'platform': true,
    'shortName': 'a shortname',
    'description': 'Some description',
    'category': 'Toto',
    'order': 1,
  },
];

describe('quarkus-project', () => {
  it('parseProjectInQuery correctly', () => {
    const queryParams = parse('g=org.test&a=code-test&v=1.0.0-SNAPSHOT&b=MAVEN&c=org.toto&e=some-id&e=arc&e=quarkus-camel-netty4-http');
    const parsedProject = parseProjectInQuery(queryParams);
    expect(parsedProject).toMatchSnapshot('no-github');
  });

  it('parseProjectInQuery correctly with github', () => {
    const queryParams = parse('g=org.test&a=code-test&v=1.0.0-SNAPSHOT&b=GRADLE&c=org.toto&e=some-id&e=arc&e=quarkus-camel-netty4-http&code=totototo&state=djdjdj&github=true');
    const parsedProject = parseProjectInQuery(queryParams);
    expect(parsedProject).toMatchSnapshot('github');
  });

  it('parseProjectInQuery with some other metadata', () => {
    const parsedProject = parseProjectInQuery(parse('v=1.0'));
    expect(parsedProject).toMatchSnapshot('v=1.0');
  });

  it('map extensions', () => {
    const mappedExtensions = mapExtensions(entries, [ 'arc', 'quarkus-camel-netty4-http' ]);
    expect(mappedExtensions).toMatchSnapshot('mappedExtensions');
  });

  it('toShortcut should work properly', () => {
    expect(toShortcut('io.quarkus:quarkus-my-ext')).toBe('my-ext');
    expect(toShortcut('quarkus-my-ext')).toBe('my-ext');
    expect(toShortcut('my-quarkus-ext')).toBe('my-quarkus-ext');
    expect(toShortcut('io.quarkiverse.myext:quarkus-my-ext')).toBe('io.quarkiverse.myext:quarkus-my-ext');
    expect(toShortcut('org.apache.camel.quarkus:camel-quarkus-core')).toBe('org.apache.camel.quarkus:camel-quarkus-core');
  });

  it('retrieveProjectFromLocalStorage should load app config from localstorage', () => {
    retrieveProjectFromLocalStorage();

    expect(localStorage.getItem).toHaveBeenCalledWith(LocalStorageKey.DEFAULT_PROJECT);
  });

  it('saveProjectToLocalStorage should save app config to localstorage', () => {
    const quarkusProject = parseProjectInQuery(parse('g=org.test&a=code-test&e=arc')) as QuarkusProject;
    const jsonProject = JSON.stringify(quarkusProject);

    saveProjectToLocalStorage(quarkusProject);

    expect(localStorage.setItem).toHaveBeenCalledWith(LocalStorageKey.DEFAULT_PROJECT, jsonProject);
  });

  it('resetProjectToDefault delete app config to localstorage', () => {
    resetProjectToDefault();

    expect(localStorage.removeItem).toHaveBeenCalledWith(LocalStorageKey.DEFAULT_PROJECT);
  });


  it('existsStoredProject should search for app config on localstorage', () => {
    existsStoredProject();

    expect(localStorage.getItem).toHaveBeenCalledWith(LocalStorageKey.DEFAULT_PROJECT);
  });
  
});