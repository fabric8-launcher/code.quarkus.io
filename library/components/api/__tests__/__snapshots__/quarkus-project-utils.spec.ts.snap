// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quarkus-project map extensions: mappedExtensions 1`] = `
Object {
  "mapped": Array [
    Object {
      "category": "Core",
      "default": false,
      "description": "Build time CDI dependency injection",
      "id": "io.quarkus:quarkus-arc",
      "keywords": Array [
        "arc",
        "cdi",
        "dependency-injection",
        "di",
        "label",
      ],
      "name": "ArC",
      "order": 0,
      "platform": true,
      "shortName": "CDI",
      "tags": Array [],
      "version": "test-version",
    },
    Object {
      "category": "Integration",
      "default": false,
      "description": "Camel support for Netty",
      "id": "io.quarkus:quarkus-camel-netty4-http",
      "keywords": Array [
        "camel-netty4-http",
        "camel",
      ],
      "name": "Camel Netty4 test HTTP",
      "order": 2,
      "platform": true,
      "tags": Array [
        "status:preview",
      ],
      "version": "test-version",
    },
  ],
  "missing": Array [],
}
`;

exports[`quarkus-project parseProjectInQuery correctly with github: github 1`] = `
Object {
  "extensions": Array [
    "some-id",
    "arc",
    "camel-netty4-http",
  ],
  "github": Object {
    "code": "totototo",
    "state": "djdjdj",
  },
  "metadata": Object {
    "artifactId": "code-test",
    "buildTool": "GRADLE",
    "groupId": "org.test",
    "javaVersion": "17",
    "noCode": false,
    "version": "1.0.0-SNAPSHOT",
  },
  "platformOnly": undefined,
  "streamKey": undefined,
}
`;

exports[`quarkus-project parseProjectInQuery correctly: no-github 1`] = `
Object {
  "extensions": Array [
    "some-id",
    "arc",
    "camel-netty4-http",
  ],
  "github": undefined,
  "metadata": Object {
    "artifactId": "code-test",
    "buildTool": "MAVEN",
    "groupId": "org.test",
    "javaVersion": "17",
    "noCode": false,
    "version": "1.0.0-SNAPSHOT",
  },
  "platformOnly": undefined,
  "streamKey": undefined,
}
`;

exports[`quarkus-project parseProjectInQuery with some other metadata: v=1.0 1`] = `
Object {
  "extensions": Array [],
  "github": undefined,
  "metadata": Object {
    "artifactId": "code-with-quarkus",
    "buildTool": "MAVEN",
    "groupId": "org.acme",
    "javaVersion": "17",
    "noCode": false,
    "version": "1.0",
  },
  "platformOnly": undefined,
  "streamKey": undefined,
}
`;
